frontend/src/metabase/visualizations/lib/apply_tooltips.js

```diff
@@ -6,160 +6,207 @@ import d3 from "d3";
-                             data = d.key._origin.row.map((value, index) => {
-                                 const col = d.key._origin.cols[index];
-                                 return { key: getFriendlyName(col), value: value, col };
-                             });
-                         } else {
-                             data = d.key.map((value, index) => (
-                                 { key: getFriendlyName(cols[index]), value: value, col: cols[index] }
-                             ));
-                         }
-                     } else if (d.data) { // line, area, bar
-                         if (!isSingleSeriesBar) {
-                             cols = series[seriesIndex].data.cols;
-                         }
-
-                         data = [
-                             {
-                                 key: getFriendlyName(cols[0]),
-                                 value: d.data.key,
-                                 col: cols[0]
-                             },
-                             {
-                                 key: getFriendlyName(cols[1]),
-                                 value: isNormalized
-                                      ? `${formatValue(d.data.value) * 100}%`
-                                      : d.data.value,
-                                 col: cols[1]
-                             }
-                         ];
-                     }
-
-                     if (data && series.length > 1) {
-                         if (card._breakoutColumn) {
-                             data.unshift({
-                                 key: getFriendlyName(card._breakoutColumn),
-                                 value: card._breakoutValue,
-                                 col: card._breakoutColumn
-                             });
-                         }
-                     }
-
-                     data = _.uniq(data, (d) => d.col);
-
-                     onHoverChange({
-                         // for single series bar charts, fade the series and highlght the hovered element with CSS
-                         index: isSingleSeriesBar ? -1 : seriesIndex,
-                         // for area charts, use the mouse location rather than the DOM element
-                         element: isArea ? null : this,
-                         event: isArea ? d3.event : null,
-                         data: data.length > 0 ? data : null,
+function setHoverFunctions(chart, series, isStacked, isNormalized, onHoverChange) {
+    let [{ data: { cols } }] = series;
+    chart.selectAll(".bar, .dot, .area, .line, .bubble")
+         .on("mousemove", function(d, i) {
+             const seriesIndex       = determineSeriesIndexFromElement(this, isStacked);
+             const thisSeries        = series[seriesIndex];
+             const card              = thisSeries.card;
+             const isSingleSeriesBar = this.classList.contains("bar") && series.length === 1;
```

Comment:
From my perspective it'd be a little nicer to have `chartType` or similar come in from the top and rather than having to do checks for what type it is inside of here. It might also let us use some easier to parse logic for handling chart types instead of a bunch of `isType` checks in various spots. For example 

```js
switch(chartType) {
  case 'area': 
   return {.//.. areaProperties or perhaps a fn call that return them}
}
```
Not sure if it makes sense to do in this round of refactoring but in my ideal world we'd move towards that kind of structure.

Reply:
Yeah, I agree. I was trying not to do too much refactoring in a bugfix PR so I thought breaking out the 200-line function into four smaller ones was enough of a win for this PR

Overall I'm hoping I can do a continuous series of PRs with a bugfix + code cleanup and this will gradually turn into a nice library of code

Reply:
Cool. That sounds good to me. 👍